File: HW1/main.py
import pandas as pd
import numpy as np
import os
import shutil
import argparse


def normalize(features: np.array):
    """
    input: features: np.array of shape (n_samples, n_features)
    output: normalized_features: np.array of shape (n_samples, n_features)
        mean: float
        std: float
    """
    return (features - features.mean(axis=0)) / features.std(axis=0), features.mean(axis=0), features.std(axis=0)

def design_matrix_single_feature(features: np.array, M, k=2):
    '''
    input: given a features matrix of shape (n_samples, n_features)
    only get the the third feature
    output is a matrix of shape (n_samples, M)
    '''
    s = 0.1
    N = features.shape[0]
    phi = np.zeros((N, M))
    features = features[:,k]
    
    for n in range(N):
        for j in range(M):
            mu_j = 3* (-M+1 +2*(j-1)*(M-1)/(M-2+1e-9)) / M
            phi[n, j] = 1 if j == 0 else sigmoid((features[n]-mu_j)/s)
                  
    return phi

def design_matrix(features: np.array, M):
    '''
    input: given a features matrix of shape (n_samples, n_features)
    a feature is transformed into a new feature vector of shape (M, 1) by applying the basis function
    output is a matrix of shape (n_samples, n_features* M)
    '''
    s = 0.1
    N = features.shape[0]
    K = features.shape[1]
    
    # create a matrix of shape (N, K, M)
    phi = np.zeros((N, K*M))
    
    phi = np.concatenate([design_matrix_single_feature(features, M, k) for k in range(K)], axis=1)

    # for n in range(N):
    #     for k in range(K*M):
    #         f = k // M # the index of the feature
    #         j = k % M # the index of the basis function
    #         mu_j = 3* (-M+1 +2*(j-1)*(M-1)/(M-2)) / M
    #         phi[n,k] = 1 if j == 0 else sigmoid( (features[n,f]-mu_j) /s)
    
    print(phi.shape)
    return phi
    

def plot_fitting_curve(features: np.array,k: int,predictions : np.array, targets : np.array,m: int):
    '''
    input: features: np.array of shape (n_samples, n_features)
        k: int, the index of the feature to plot (=2 for danceability)
        targets: np.array of shape (n_samples, 1)
        weights: np.array of shape (M, 1)
        M: int, the number of basis functions
    '''
    
    import matplotlib.pyplot as plt
    plt.cla()
    feature = features[:,k]
    plt.scatter(feature, targets, label='true data')
    plt.scatter(feature, predictions, label='Predicted')
    # add the plot title
    plt.title('Fitting curve for M =  {} at test data'.format(m))
    plt.legend()
    plt.savefig(os.path.join('plot','M_{}.png'.format(m)))
    

def generate(train_features: np.array, test_features: np.array, train_targets: np.array, M: int, lambda_: float = 0.1):
    '''
    input:  train_features: np.array of shape (n_samples, n_features)
            test_features: np.array of shape (n_samples, n_features)
            train_targets: np.array of shape (n_samples, 1)
            M: int, the number of basis functions
            lambda_: float, the regularization parameter
    output: train_phi: np.array of shape (n_samples, K*M)
            test_phi: np.array of shape (n_samples, K*M)
            weights: np.array of shape (K*M, 1)
            train_predictions: np.array of shape (n_samples, 1)
            test_predictions: np.array of shape (n_samples, 1)
    '''
    
    train_phi = design_matrix(train_features, M=M)
    test_phi = design_matrix(test_features, M=M)
    
    print('train_phi has shape {}'.format(train_phi.shape)) # N, K*M
    print('test_phi has shape {}'.format(test_phi.shape)) # N, K*M
    if lambda_ == 0:
        weights = np.linalg.pinv(train_phi) @ train_targets
    else:
        weights = np.linalg.inv(train_phi.T @ train_phi + lambda_ * np.eye(M*train_features.shape[1])) @ train_phi.T @ train_targets
    
    print('weights has shape {}'.format(weights.shape)) # K*M, 1
    
    train_predictions = train_phi @ weights
    test_predictions = test_phi @ weights
    
    return train_phi, test_phi, weights, train_predictions, test_predictions

def mse(predictions: np.array, targets: np.array):
    '''
    input: predictions: np.array of shape (n_samples, 1)
           targets: np.array of shape (n_samples, 1)
    output: mse: float
    '''
    return np.mean((predictions - targets)**2)

def accuracy(predictions: np.array, targets: np.array):
    # modify the test_target is that the 0 becomes 1
    targets = [targets[i] if targets[i] != 0 else 1 for i in range(len(targets))]
    accuracy = 1 - np.mean(np.abs((predictions - targets)/targets))
    return accuracy

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def plot_error(train_error: np.array, test_error: np.array):
    import matplotlib.pyplot as plt
    plt.cla()
    plt.plot(M, train_error, label='train error')
    plt.plot(M, test_error, label='test error')
    plt.legend()
    plt.title('Train and Test Error')
    plt.savefig('error.png')

def plot_accuracy(train_accuracy: np.array, test_accuracy: np.array):
    import matplotlib.pyplot as plt
    plt.cla()
    plt.plot(M, train_accuracy, label='train accuracy')
    plt.plot(M, test_accuracy, label='test accuracy')
    plt.legend()
    plt.title('Train and Test Accuracy')
    plt.savefig('accuracy.png')

def plot_weights(weights: np.array,m: int):
    import matplotlib.pyplot as plt
    plt.cla()
    '''
    the weights are of shape (K*M, 1)
    for each feature, plot the weights of the basis functions
    '''
    sum = []
    for k in range(11):
        sum.append(np.abs(weights[k*m:(k+1)*m]).sum())
    plt.bar(range(11), sum)
    plt.title('Sum of weights for each feature at M = {}'.format(m))
    plt.savefig(f'weights at {m}.png')
    
    
if __name__ == "__main__":
    
    parser = argparse.ArgumentParser()
    parser.add_argument('--lambda_', type=float, default=0.1)
    args = parser.parse_args()
    df = pd.read_csv('HW1.csv')
    targets = np.array(df['song_popularity'])
    features = np.array(df.drop('song_popularity', axis=1))

    # split into testing and training data
    train_features = features[:10000]
    test_features = features[10000:]
    train_targets = targets[:10000]
    test_targets = targets[10000:]
    
    print('train_features has {} samples and {} features'.format(*train_features.shape))
    print('train_targets has {} samples'.format(*train_targets.shape))
    print('test_features has {} samples and {} features'.format(*test_features.shape))
    print('test_targets has {} samples'.format(*test_targets.shape))
    
    # normalize the features
    normalized_train_features, mean, std = normalize(train_features)
    normalized_test_features = (test_features - mean) / std
    
    # ============================================ #
    if os.path.exists('plot'):
        shutil.rmtree('plot')
    os.mkdir('plot')
    
    train_error = []
    test_error = []
    train_accuracy = []
    test_accuracy = []
    M = [5,10,15,20,25,30]
    for m in M:
        print('M = {} .......'.format(m))
        train_phi, test_phi, weights, train_predictions, test_predictions = generate(normalized_train_features, normalized_test_features,train_targets,M=m,lambda_= args.lambda_)
        
        # compute the error
        plot_fitting_curve(normalized_test_features, 2, test_predictions, test_targets, m=4)
        plot_weights(weights,m)
        train_error.append(mse(train_predictions, train_targets))
        test_error.append(mse(test_predictions, test_targets))
        train_accuracy.append(accuracy(train_predictions, train_targets))
        test_accuracy.append(accuracy(test_predictions, test_targets))
        
    plot_error(train_error, test_error)
    plot_accuracy(train_accuracy, test_accuracy)

File: HW1/main.cpp
# include <iostream>
# include <fstream>
# include <string>  
# include <sstream>
# include <eigen3/Eigen/Dense>
using namespace std;

double sigmoid(double x) {
    return 1 / (1 + exp(-x));
}

Eigen::MatrixXd design_matrix(Eigen::MatrixXd features, int M) {
    double s = 0.1;
    int N = features.rows();
    int K = features.cols();
    Eigen::MatrixXd phi = Eigen::MatrixXd::Zero(N, M * K);
    for(int n = 0; n < N; n++) {
        for(int k = 0; k < K * M; k++) {
            int f = k / M; // the index of the feature
            int j = k % M; // the index of the basis function
            double mu_j = 3 * (-M + 1 + 2 * (j - 1) * (M - 1) / (M - 2 +1e-9)) / M;
            phi(n, k) = j == 0 ? 1 : sigmoid((features(n, f) - mu_j) / s);
        }
    }
    // std::cout << phi.rows() << ", " << phi.cols() << std::endl;
    return phi;
}

int countRowsInCSV(const std::string& filename) {
    std::ifstream file(filename);
    std::string line;
    int count = 0;

    while (std::getline(file, line)) {
        ++count;
    }

    return count  - 1;
}

int main(){
    // read data ***********************//
    // read file from a csv file
    std::ifstream file;
    file.open("HW1.csv");
    std::string line;
    
    // ***************demo***************//
    ifstream demo_file;
    demo_file.open("HW1_demo.csv");
    //******************************//

    // Read and discard the header line
    std::getline(file, line);
    Eigen::MatrixXd targets(15817, 1);
    Eigen::MatrixXd features(15817, 11);


    int row = 0;
    while(std::getline(file, line)){
        std::stringstream ss(line);
        std::string data;
        int col = 0;
        while(std::getline(ss, data, ',')){
            if (col == 0) { // assuming the first column is the target
                targets(row, 0) = std::stod(data);
            } else { // the rest are features
                features(row, col-1) = std::stod(data);
            }
            col++;
        }
        row++;
    }
    file.close();

    //*************demo******************//
    // Read and discard the header line
    std::getline(demo_file, line);
    int demo_count = countRowsInCSV("HW1.csv");
    Eigen::MatrixXd demo_features(demo_count, 11);
    Eigen::MatrixXd demo_targets(demo_count, 1);

    row = 0;
    while(std::getline(demo_file, line)){
        std::stringstream ss(line);
        std::string data;
        int col = 0;
        while(std::getline(ss, data, ',')){
            if (col == 0) { // assuming the first column is the target
                demo_targets(row, 0) = std::stod(data);
            } else { // the rest are features
                demo_features(row, col-1) = std::stod(data);
            }
            col++;
        }
        row++;
    }
    demo_file.close();
    //*******************************//

    // Assuming features and targets are Eigen::MatrixXd
    Eigen::MatrixXd train_features = Eigen::MatrixXd::Zero(10000, 11);
    Eigen::MatrixXd train_targets = Eigen::MatrixXd::Zero(10000, 1);
    Eigen::MatrixXd test_features = Eigen::MatrixXd::Zero(5817, 11);
    Eigen::MatrixXd test_targets = Eigen::MatrixXd::Zero(5817, 1);
    
    for(int i = 0; i < 10000; i++) {
        for(int j = 0; j < 11; j++) {
            train_features(i, j) = features(i, j);
        }
        train_targets(i, 0) = targets(i, 0);
    }

    for(int i = 10000; i < 15817; i++) {
        for(int j = 0; j < 11; j++) {
            test_features(i - 10000, j) = features(i, j);
        }
        test_targets(i - 10000, 0) = targets(i, 0);
    }


    // Assuming train_features and test_features are Eigen::MatrixXd
    Eigen::VectorXd mean = Eigen::VectorXd::Zero(11);
    Eigen::VectorXd std = Eigen::VectorXd::Zero(11);

    for(int i = 0; i < 11; i++) {
        double sum = train_features.col(i).sum();
        mean(i) = sum / 10000.;
        double sum_squared = (train_features.col(i).array() - mean(i)).square().sum();
        std(i) = sqrt(sum_squared / 10000); // Use Bessel's correction
    }

    for(int i = 0; i < 10000; i++) {
        for(int j = 0; j < 11; j++) {
            train_features(i, j) = (train_features(i, j) - mean(j)) / std(j);
        }
    }
    for(int i = 0; i < 5817; i++) {
        for(int j = 0; j < 11; j++) {
            test_features(i, j) = (test_features(i, j) - mean(j)) / std(j);
        }
    }


    //***************demo***************//
    for (int i = 0; i < demo_count; i++) {
        for (int j = 0; j < 11; j++) {
            demo_features(i, j) = (demo_features(i, j) - mean(j)) / std(j);
        }
    }
    //***********************************//

    double lambda_ = 0.1;
    double train_error; 
    double test_error;
    double train_error_regularization; 
    double test_error_regularization;
    double train_accuracy;
    double test_accuracy;
    double train_accuracy_regularization;
    double test_accuracy_regularization;
    ofstream outfile("result.csv");
    outfile << "M,Train Error,Train Accuracy,Test Error,Test Accuracy,Train Error Regularization,Train Accuracy Regularization,Test Error Regularization,Test Accuracy Regularization\n";
    ofstream outfile5("demo_result.csv");
    outfile5 << "M,Demo Accuracy, Demo Accuracy Regularization\n";
    ofstream outfile3("cross_validation.csv");
    outfile3 << "M,Validation Error\n";
    for (int M=1; M<=30; M++) {
        if (M == 5 || M == 10 || M == 15 || M == 20 || M == 25 || M == 30) {
            cout << "M: " << M << endl;
            Eigen::MatrixXd train_phi = design_matrix(train_features, M);
            Eigen::MatrixXd test_phi = design_matrix(test_features, M);
            Eigen::MatrixXd w = train_phi.bdcSvd(Eigen::ComputeThinU | Eigen::ComputeThinV).solve(train_targets);
            Eigen::MatrixXd I = Eigen::MatrixXd::Identity(M * train_features.cols(), M * train_features.cols());
            Eigen::MatrixXd w2 = (train_phi.transpose() * train_phi + lambda_ * I).ldlt().solve(train_phi.transpose() * train_targets);
            Eigen::MatrixXd train_predictions = train_phi * w;
            Eigen::MatrixXd test_predictions = test_phi * w;
            Eigen::MatrixXd train_predictions_regularization = train_phi * w2;
            Eigen::MatrixXd test_predictions_regularization = test_phi * w2;
            train_error = (train_predictions - train_targets).array().square().sum() / 10000.;
            test_error = (test_predictions - test_targets).array().square().sum() / 5817.;
            train_error_regularization = (train_predictions_regularization - train_targets).array().square().sum() / 10000.;
            test_error_regularization = (test_predictions_regularization - test_targets).array().square().sum() / 5817.;
            
            //************demo**********//
            // plot the fitting curve of the third feature of the demo set
            Eigen::MatrixXd demo_phi = design_matrix(demo_features, M);
            Eigen::MatrixXd demo_predictions = demo_phi * w;
            Eigen::MatrixXd demo_predictions_regularization = demo_phi * w2;
            ofstream outfile4("demo_predictions_" + std::to_string(M) + ".csv");
            outfile4 << "danceability,Predictions,Predictions Regularization,Test_target\n";
            for (int i = 0; i < demo_count; i++) {
                outfile4 << demo_features(i, 2) << "," << demo_predictions(i, 0) << "," << demo_predictions_regularization(i, 0) << "," << demo_targets(i, 0) << "\n";
                outfile4.flush();
            }
            outfile4.close();
            //**************************//

            
            // create a target_temp to calculate the accuracy
            Eigen::MatrixXd train_targets_temp = train_targets;
            Eigen::MatrixXd test_targets_temp = test_targets;

            for(int i = 0; i < 10000; i++) {
                if(train_targets_temp(i, 0) == 0) {
                    train_targets_temp(i, 0) = 1;
                }
            }
            for(int i = 0; i < 5817; i++) {
                if(test_targets_temp(i, 0) == 0) {
                    test_targets_temp(i, 0) = 1;
                }
            }

            // ************demo************//
            // create a target_temp to calculate the accuracy
            Eigen::MatrixXd demo_targets_temp = demo_targets;
            for (int i = 0; i < demo_count; i++) {
                if (demo_targets_temp(i, 0) == 0) {
                    demo_targets_temp(i, 0) = 1;
                }
            }
            //*****************************//

            
            Eigen::MatrixXd train_diff = Eigen::MatrixXd::Zero(10000, 1);
            for (int i = 0; i < 10000; i++) {
                train_diff(i, 0) = abs((train_predictions(i, 0) - train_targets_temp(i, 0)) / train_targets_temp(i, 0));
            }
            train_accuracy = 1 - train_diff.sum() / 10000.;

            Eigen::MatrixXd test_diff = Eigen::MatrixXd::Zero(5817, 1);
            for (int i = 0; i < 5817; i++) {
                test_diff(i, 0) = abs((test_predictions(i, 0) - test_targets_temp(i, 0)) / test_targets_temp(i, 0));
            }
            test_accuracy = 1 - test_diff.sum() / 5817.;

            Eigen::MatrixXd train_diff_regularization = Eigen::MatrixXd::Zero(10000, 1);
            for (int i = 0; i < 10000; i++) {
                train_diff_regularization(i, 0) = abs((train_predictions_regularization(i, 0) - train_targets_temp(i, 0)) / train_targets_temp(i, 0));
            }
            train_accuracy_regularization = 1 - train_diff_regularization.sum() / 10000.;

            Eigen::MatrixXd test_diff_regularization = Eigen::MatrixXd::Zero(5817, 1);
            for (int i = 0; i < 5817; i++) {
                test_diff_regularization(i, 0) = abs((test_predictions_regularization(i, 0) - test_targets_temp(i, 0)) / test_targets_temp(i, 0));
            }
            test_accuracy_regularization = 1 - test_diff_regularization.sum() / 5817.;

            outfile << M << "," << train_error << "," << train_accuracy << "," << test_error << "," << test_accuracy << "," << train_error_regularization << "," << train_accuracy_regularization << "," << test_error_regularization << "," << test_accuracy_regularization << "\n";
            outfile.flush();
            cout << M << "," << train_error << "," << train_accuracy << "," << test_error << "," << test_accuracy << "," << train_error_regularization << "," << train_accuracy_regularization << "," << test_error_regularization << "," << test_accuracy_regularization << "\n";
            
            // ***********demo*****************//
            // calculate the accuracy of the demo set
            Eigen::MatrixXd demo_diff = Eigen::MatrixXd::Zero(demo_count, 1);
            Eigen::MatrixXd demo_diff_regularization = Eigen::MatrixXd::Zero(demo_count, 1);
            for (int i = 0; i < demo_count; i++) {
                demo_diff(i, 0) = abs((demo_predictions(i, 0) - demo_targets_temp(i, 0)) / demo_targets_temp(i, 0));
                demo_diff_regularization(i, 0) = abs((demo_predictions_regularization(i, 0) - demo_targets_temp(i, 0)) / demo_targets_temp(i, 0));
            }
            double demo_accuracy = 1 - demo_diff.sum() / demo_count;
            double demo_accuracy_regularization = 1 - demo_diff_regularization.sum() / demo_count;
            cout << "Demo Accuracy: " << demo_accuracy << ", Demo Accuracy Regularization: " << demo_accuracy_regularization << endl;
            outfile5 << M << "," << demo_accuracy << "," << demo_accuracy_regularization << "\n";
            outfile5.flush();
            //*********************************//
            
            
            //  plot the fitting curve of the third feature 
            string filename = "fitting_curve_" + std::to_string(M) + ".csv";
            ofstream outfile2(filename);
            outfile2 << "danceability,Predictions,Predictions Regularization,Test_target\n";
            // test set
            // for (int i = 0; i < 5817 ; i++) {
            //     outfile2 <<  test_features(i,2) << "," << test_predictions(i, 0) << "," << test_predictions_regularization(i, 0) << "," << test_targets(i,0)<< "\n";
            //     outfile2.flush();
            // }
            // train set
            for (int i = 0; i < 10000 ; i++) {
                outfile2 <<  train_features(i,2) << "," << train_predictions(i, 0) << "," << train_predictions_regularization(i, 0) << "," << train_targets(i,0)<< "\n";
                outfile2.flush();
            }
            outfile2.close();

            // solution 2
            // Eigen::MatrixXd curve_features = Eigen::MatrixXd::Zero(10000, 11);
            // double min = 0;
            // double max = 1;
            // for (int i = 0; i < 10000; i++) {
            //     curve_features(i,2) = i / 10000.;
            // }
            

            // Eigen::MatrixXd phi = design_matrix(curve_features, M);
            // Eigen::MatrixXd predictions = phi * w;
            // Eigen::MatrixXd predictions_regularization = phi * w2;
            // for (int i = 0; i < 10000; i++) {
            //     outfile2 << curve_features(i, 2) << "," << predictions(i, 0) << "," << predictions_regularization(i, 0) << "\n";
            // }
            // outfile2.close();

        }
        
        // *************************************************************************************//
        
        if (1) {
            ///////// apply five-fold cross-validation /////////////////////////
            Eigen::MatrixXd cross_train_features = Eigen::MatrixXd::Zero(8000, 11);
            Eigen::MatrixXd cross_train_targets = Eigen::MatrixXd::Zero(8000, 1);
            Eigen::MatrixXd cross_validation_features = Eigen::MatrixXd::Zero(2000, 11);
            Eigen::MatrixXd cross_validation_targets = Eigen::MatrixXd::Zero(2000, 1);
            double valid_error = 0;

            // first fold (0-8000, 8000-10000) ////////
            for(int i = 0; i < 8000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_train_features(i, j) = train_features(i, j);
                }
                cross_train_targets(i, 0) = train_targets(i, 0);
            }
            for(int i = 8000; i < 10000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_validation_features(i - 8000, j) = train_features(i, j);
                }
                cross_validation_targets(i - 8000, 0) = train_targets(i, 0);
            }

            Eigen::MatrixXd cross_train_phi = design_matrix(cross_train_features, M);
            Eigen::MatrixXd cross_valid_phi = design_matrix(cross_validation_features, M);
            Eigen::MatrixXd cross_w = cross_train_phi.bdcSvd(Eigen::ComputeThinU | Eigen::ComputeThinV).solve(cross_train_targets);
            Eigen::MatrixXd cross_valid_predictions = cross_valid_phi * cross_w;
            valid_error += (cross_valid_predictions - cross_validation_targets).array().square().sum() / 2000.;

            // second fold (0-6000, 6000-8000, 8000-10000)
            for(int i = 0; i < 6000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_train_features(i, j) = train_features(i, j);
                }
                cross_train_targets(i, 0) = train_targets(i, 0);
            }
            for(int i = 6000; i < 8000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_validation_features(i - 6000, j) = train_features(i, j);
                }
                cross_validation_targets(i - 6000, 0) = train_targets(i, 0);
            }
            for(int i = 8000; i < 10000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_train_features(i-2000, j) = train_features(i, j);
                }
                cross_train_targets(i-2000, 0) = targets(i, 0);
            }

            cross_train_phi = design_matrix(cross_train_features, M);
            cross_valid_phi = design_matrix(cross_validation_features, M);
            cross_w = cross_train_phi.bdcSvd(Eigen::ComputeThinU | Eigen::ComputeThinV).solve(cross_train_targets);
            cross_valid_predictions = cross_valid_phi * cross_w;
            valid_error += (cross_valid_predictions - cross_validation_targets).array().square().sum() / 2000.;

            // third fold (0-4000, 4000-6000, 6000-10000)
             for(int i = 0; i < 4000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_train_features(i, j) = train_features(i, j);
                }
                cross_train_targets(i, 0) = train_targets(i, 0);
            }
            for(int i = 4000; i < 6000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_validation_features(i-4000, j) = train_features(i, j);
                }
                cross_validation_targets(i - 4000, 0) = train_targets(i, 0);
            }
            for(int i = 6000; i < 10000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_train_features(i-2000, j) = train_features(i, j);
                }
                cross_train_targets(i-2000, 0) = targets(i, 0);
            }

            cross_train_phi = design_matrix(cross_train_features, M);
            cross_valid_phi = design_matrix(cross_validation_features, M);
            cross_w = cross_train_phi.bdcSvd(Eigen::ComputeThinU | Eigen::ComputeThinV).solve(cross_train_targets);
            cross_valid_predictions = cross_valid_phi * cross_w;
            valid_error += (cross_valid_predictions - cross_validation_targets).array().square().sum() / 2000.;
            

            // fourth fold (0-2000,2000-4000, 4000-10000)
            for(int i = 0; i < 2000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_train_features(i, j) = train_features(i, j);
                }
                cross_train_targets(i, 0) = train_targets(i, 0);
            }
            for(int i = 2000; i < 4000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_validation_features(i - 2000, j) = train_features(i, j);
                }
                cross_validation_targets(i - 2000, 0) = train_targets(i, 0);
            }
            for(int i = 6000; i < 10000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_train_features(i-2000, j) = train_features(i, j);
                }
                cross_train_targets(i-2000, 0) = train_targets(i, 0);
            }
            
            cross_train_phi = design_matrix(cross_train_features, M);
            cross_valid_phi = design_matrix(cross_validation_features, M);
            cross_w = cross_train_phi.bdcSvd(Eigen::ComputeThinU | Eigen::ComputeThinV).solve(cross_train_targets);
            cross_valid_predictions = cross_valid_phi * cross_w;
            valid_error += (cross_valid_predictions - cross_validation_targets).array().square().sum() / 2000.;
        

            // fifth fold (0-2000,2000-10000)
            for(int i = 0; i < 2000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_validation_features(i, j) = train_features(i, j);
                }
                cross_train_targets(i, 0) = targets(i, 0);
            }
            for(int i = 2000; i < 10000; i++) {
                for(int j = 0; j < 11; j++) {
                    cross_train_features(i-2000, j) = train_features(i, j);
                }
                cross_train_targets(i-2000, 0) = train_targets(i, 0);
            }

            cross_train_phi = design_matrix(cross_train_features, M);
            cross_valid_phi = design_matrix(cross_validation_features, M);
            cross_w = cross_train_phi.bdcSvd(Eigen::ComputeThinU | Eigen::ComputeThinV).solve(cross_train_targets);
            cross_valid_predictions = cross_valid_phi * cross_w;
            valid_error += (cross_valid_predictions - cross_validation_targets).array().square().sum() / 2000.;

            valid_error /= 5;
            cout<< M << " " << valid_error << endl;
            outfile3 << M << "," << valid_error << "\n";
        } // end if
    } // end for
    outfile.close();
    outfile3.close();
    outfile5.close();
    return 0;
}
